contracts/EIP712CrossChain.sol

```diff
@@ -0,0 +1,81 @@
+// SPDX-License-Identifier: GPL-3.0-only
+pragma solidity ^0.8.0;
+
+import "@openzeppelin/contracts/utils/cryptography/ECDSA.sol";
+
+/**
+ * @dev https://eips.ethereum.org/EIPS/eip-712[EIP 712] is a standard for hashing and signing of typed structured data.
+ *
+ * This contract is based on OpenZeppelin's implementation:
+ * https://github.com/OpenZeppelin/openzeppelin-contracts/blob/v4.8.0/contracts/utils/cryptography/EIP712.sol
+ *
+ * NOTE: Modified version that allows to build a domain separator that relies on a different chain id than the chain this
+ * contract is deployed to. An example use case we want to support is:
+ * - User A signs a message on chain with id = 1
+ * - User B executes a method by verifying user A's EIP-712 compliant signature on a chain with id != 1
+ */
+abstract contract EIP712CrossChain {
+    /* solhint-disable var-name-mixedcase */
+    bytes32 private immutable _HASHED_NAME;
+    bytes32 private immutable _HASHED_VERSION;
+    bytes32 private immutable _TYPE_HASH;
+
+    /* solhint-enable var-name-mixedcase */
+
+    /**
+     * @dev Initializes the domain separator and parameter caches.
+     *
+     * The meaning of `name` and `version` is specified in
+     * https://eips.ethereum.org/EIPS/eip-712#definition-of-domainseparator[EIP 712]:
+     *
+     * - `name`: the user readable name of the signing domain, i.e. the name of the DApp or the protocol.
+     * - `version`: the current major version of the signing domain.
+     *
+     * NOTE: These parameters cannot be changed except through a xref:learn::upgrading-smart-contracts.adoc[smart
+     * contract upgrade].
+     */
+    constructor(string memory name, string memory version) {
+        bytes32 hashedName = keccak256(bytes(name));
+        bytes32 hashedVersion = keccak256(bytes(version));
+        bytes32 typeHash = keccak256("EIP712Domain(string name,string version,uint256 chainId)");
+        _HASHED_NAME = hashedName;
+        _HASHED_VERSION = hashedVersion;
+        _TYPE_HASH = typeHash;
+    }
+
+    /**
+     * @dev Returns the domain separator depending on the `originChainId`.
+     */
+    function _domainSeparatorV4(uint256 originChainId) internal view returns (bytes32) {
+        return _buildDomainSeparator(_TYPE_HASH, _HASHED_NAME, _HASHED_VERSION, originChainId);
+    }
+
+    function _buildDomainSeparator(
+        bytes32 typeHash,
+        bytes32 nameHash,
+        bytes32 versionHash,
+        uint256 originChainId
+    ) private pure returns (bytes32) {
+        return keccak256(abi.encode(typeHash, nameHash, versionHash, originChainId));
+    }
```

Comment:
nit: can we combine these methods to simplify since we're not intending to use _buildDomainSeparator outside of _domainSeparatorV4?

Reply:
Done here https://github.com/across-protocol/contracts-v2/pull/222/commits/21aab9c2453ce1a06c4833ef7cf3e77a1d2b9d1f